src/OpenLoco/Input/Keyboard.cpp

```diff
@@ -142,11 +146,42 @@ namespace OpenLoco::Input
         }
     }
 
+    // 0x00406FBA
+    void enqueueKey(uint32_t keycode)
+    {
+        uint32_t writeIndex = _keyQueueWriteIndex;
+        auto nextWriteIndex = (writeIndex + 1) & (64 - 1);
+        if (nextWriteIndex == _keyQueueReadIndex)
+        {
+            return;
+        }
+        _keyQueueLastWrite = writeIndex;
+        _keyQueue[writeIndex] = { keycode, 0 };
+        _keyQueueWriteIndex = nextWriteIndex;
+    }
+
+    void enqueueText(const char* text)
+    {
+        if (text != nullptr && text[0] != '\0')
+        {
+            auto index = _keyQueueLastWrite;
+            _keyQueue[index].charCode = text[0];
+        }
+    }
+
     static key* getNextKey()
     {
-        registers regs;
-        call(0x00407028, regs);
-        return (key*)regs.eax;
+        uint32_t readIndex = _keyQueueReadIndex;
+        if (readIndex == _keyQueueWriteIndex)
+        {
+            return nullptr;
+        }
+        auto* out = &_keyQueue[readIndex];
+        readIndex++;
+        // Wrap around at _keyQueue size
+        readIndex &= (64 - 1);
```

Comment:
Can we use something like
```suggestion
        readIndex &= std::size(_keyQueue) - 1;
```
or indeed a modulo operator?